### ISSUE TITLE
Redistribute disk usage on `atd-data02` to give `docker` additional headroom

### ISSUE THREAD

**Issue by frankhereford** [cityofaustin/atd-data-tech#9299]
## TLDR

We've run out of disk space for docker on `atd-data02`, and we need to juggle some data around on the server to free up some headroom for starting new containers. We have bought ourselves a day or two of time to handle the situation, but a permanent fix is required to maintain our `airflow` service functionality. 

## Request

I would like to invite all members of the team to review my proposal below, and I would like to ask our team leadership (@mateoclarke & @amenity) to also to consider the urgency and timing of when this (or another) plan may be put into action.

## Issue

`atd-data02` has 150 GB of raw disk space which is distributed over a number of filesystems using the Linux Volume Manager (LVM2) system. The distribution of that disk space between the various filesystem does not align to the usage patterns of the filesystems, and we have effectively run out of space in `/var`. The `docker` service on `atd-data02` became unable to allocate space for spinning up new containers which caused `airflow` to begin to experience errors, both internally to the `airflow` service and its children DAGs. 

## Details

`df -h | grep -v snapd | grep -v tmpfs | grep -v boot` returns the following:

```
[herefordf@atd-data02 ~]$ df -h | grep -v snapd | grep -v tmpfs | grep -v boot
Filesystem                   Size  Used Avail Use% Mounted on
/dev/mapper/root_vg-root_lv  103G   11G   93G  11% /
/dev/mapper/root_vg-home_lv  3.5G  123M  3.4G   4% /home
/dev/mapper/root_vg-var_lv    22G   22G  711M  97% /var
/dev/mapper/root_vg-tmp_lv    17G   47M   17G   1% /tmp
[herefordf@atd-data02 ~]$
```
The `/var` mount is very close to out of space. In triaging the issue earlier with @chiaberry and @johnclary, I did a `docker system prune` which cleaned up intermediate build "layers" of the images we have on the machine, which bought us a few hundred megabytes of breathing room and appears to have resolved the issue for the time being. We only have 711 megabytes of free space though, and I would expect that to become exhausted somewhere in the order of a day or two. While more small, stop-gap actions can be taken to buy us a few more days, a larger solution to give `/var` more space is required for ongoing success with `airflow`. An example stop-gap action would be to remove old logs from `/var/log`.

I have run `ncdu` in `/var` and there are no other large collections of data on disk that we can prune; docker is the sole consumer of `/var` on this machine.

I am actively monitoring the usage of `/var` today to catch it before it gets totally full with this command: `while sleep 5; do df -h | grep var | grep -v lib; done` Edit: I will continue to monitor the capacity of the file system until this is resolved; I'll keep it running. 

`lvscan` returns the following:
```
[root@atd-data02 herefordf]# lvscan
  ACTIVE            '/dev/root_vg/var_lv' [21.98 GiB] inherit
  ACTIVE            '/dev/root_vg/tmp_lv' [16.95 GiB] inherit
  ACTIVE            '/dev/root_vg/swap_lv' [4.00 GiB] inherit
  ACTIVE            '/dev/root_vg/home_lv' [<3.49 GiB] inherit
  ACTIVE            '/dev/root_vg/root_lv' [<102.99 GiB] inherit
[root@atd-data02 herefordf]#
```
You can see here that we have a bit over 100 gigabytes of space allocated to the root filesystem, `/`, which is very generous. My experience is that, because this filesystem is generally consumed by `/usr`, which contains system binaries and libraries, `/` needs to have ~20 gigabytes allocated to it and can be expected to grow exceptionally slowly. Things which grow `/usr` are things like installing new system packages which happens almost never.

You can also see that `/var` is allocated only 22 gigabytes. `/var` is the place where linux systems store its services' files on disk that contain the state and "working files" of the service. An example of this is that `/var/lib/docker` is where docker keeps all of its disk state, such as images and persistent volumes. Running out of space on this filesystem is the root of our current issue.

The server `atd-data02` is an installation of Centos 7. This flavor of linux uses the `XFS` filesystem as opposed to other usual suspects like `ext4`, `ZFS`, and `btrfs` to name a few. XFS is great filesystem: rock solid and well supported, but one feature that it does not have, in any shape or form, is the ability to shrink the filesystem. This is not uncommon, and I do not consider it a deficiency. However, that reality prevents us from shrinking the file system on `/` to free up space in our LVM volume group that could then be allocated to the `/var` logical volume. If that were possible, we'd do that and then grow the XFS filesystem for `/var` and this would be squared up. 

But we can't do that, so we need to come up with something a little bit more creative. 

## Proposal

1) We schedule 30 minutes of downtime for `airflow` and begin the following operation
1) We stop the `airflow` docker container and its related containers. (`nginx`, `postgres`, & `redis`)
1) We stop the `docker` service using `systemctl`
1) We create a directory somewhere on the `/` filesystem to become our new docker workspace. The new `/var/lib/docker` if you will. I propose `/docker`.
1) We `tar cv` (tape archive) up the docker folder (`/var/lib/docker`) to `STDOUT` so that we capture all the extended attributes (file permissions, ownership, etc) stored by XFS in its role as a POSIX filesystem.
1) We pipe that stream of data out of `tar` into a `tee` invocation that writes it to disk for archival purposes and echos it again back out on `tee`'s STDOUT
1) We pipe that stream of data into `tar xv` into the new docker home, `/docker`, so that it now contains a perfect copy of `/var/lib/docker`. 
1) We exercise appropriate paranoia and do a `rsync -avnP --delete /var/lib/docker /docker` (`-n` means dry run) to observe that we do have a perfect copy. 
1) We note the file permissions and ownership of the `/var/lib/docker` directory itself. We make sure that `/docker` mirrors those permissions and ownership.
1) We take a deep breath and take stock of our plan one last time as everything prior can be undone, no-harm-no-foul.
1) We delete `/var/lib/docker`.
1) We `ln -sd /docker /var/lib/docker` to write a cross-filesystem symlink of the new `docker` directory to show up where the old one was. 
1) We `cd /var/lib/docker` and observe that everything looks just like it did before, and that the `docker` service will be none-the-wiser that we moved its house while it was off.
1) We use `systemctl` to start the docker service and observe that `docker ps` returns no running containers and not a complaint about being unable to connect to its file socket.
1) We restart `airflow` as per the usual instructions we use when CTM reboots this machine as part of its regular service.

## Final notes
There are as many ways to solve this problem as people we might ask, so this plan is just a starting point. I am eager to discuss and revise it, and I also want to invite the whole dev & product team (@amenity @patrickm02L @mateoclarke @chiaberry @mddilley @Charlie-Henry @tillyw @johnclary) to be a part of this operation as observers and contributors. This is a fun bit of system administration and a very interesting problem we get to solve.

**Comment by Charlie-Henry** [comment]
Does this affect our Prefect instance on `atd-data02`? 

One issue may be that my parking data flows are downloading emails to disk and not deleting them after the task is complete which will slowly fill up our disk.

https://github.com/cityofaustin/atd-parking-data/blob/production/meters/fiserv_email_pub.py

**Comment by frankhereford** [comment]
@Charlie-Henry wrote:
> One issue may be that my parking data flows are downloading emails to disk and not deleting them after the task is complete which will slowly fill up our disk.

It may be! Does that flow happen in a docker image? Is docker any part of that system? 

However, even if that is the case, this server has its block device resources allocated in a sub-optimal configuration, and even if was your flow that broke the camel's back, if it hadn't been that, it would have been the next one.

**Comment by Charlie-Henry** [comment]
@frankhereford it is running in a docker image called `atddocker/atd-parking-data-meters`. I'm not sure where to look for the email files in the file system but they can all be deleted now if they are there. We have backups in S3.

**Comment by frankhereford** [comment]
@Charlie-Henry and I got to sync a second ago, and while his process had been leaving crufty volumes attached to his stopped containers, they are not a low hanging fruit we can reclaim. They had accumulated to only a few hundred megabytes, and I think they were actually a part of what we reclaimed with `docker system prune` that bought us a few days of breathing room.

**Comment by frankhereford** [comment]
## TLDR
* We ran out of space again on `/var`
* I ran `docker system prune` which this time was able to purge 12 gigabytes of data
* Everything appears to be running as expected

## Details

The `atd-data02` server ran out of space again last night. I had observed it slowly dropping over the last few days, but I expected us to still have ~300 megabytes this morning at the rate it had been going. That turns out to not be enough headroom, and the airflow container did not successfully restart. 

This morning, trying to clear up a few hundred megabytes like we did on Thursday morning, I ran `docker system prune`. This time, that command cleared up a single container, `28b77e48c44bf8b8e9ab478978065315f208d6b0795840109d84725560da3675`, but that container was a little more than 11 gigabytes. I restarted airflow, and everything came up as expected, including its past log data.

Without historical data about when this logical volume grew, it's difficult for me to know exactly what is going on here. This is exacerbated by the fact that it's reasonable to me that we were getting some unexpected behavior out of docker since we had run it out of disk space on Thursday.

That said, I happened to still have the data from our work on Thursday afternoon, and the container that existed under that hash was an airflow container. This understanding that it wasn't some _other_ container, coupled with the fact that we once again have airflow up and running without any missing data, indicates to me that we are good to close this.
